File: string/longest_substring_without_repeating_characters/longest_substring_without_repeating_characters.py
#!/usr/bin/env python3
"""
LeetCode #3: Longest Substring Without Repeating Characters

Problem: Given a string s, find the length of the longest substring 
without repeating characters.

Multiple Python implementations showcasing different approaches,
from basic to advanced optimizations and Pythonic solutions.
"""

import time
from typing import Dict, List, Tuple, Set
from collections import defaultdict, deque
import itertools

# =============================================================================
# Approach 1: Brute Force - Check all substrings
# Time: O(n^3), Space: O(min(m,n))
# =============================================================================

def lengthOfLongestSubstring_bruteforce(s: str) -> int:
    """
    Brute Force: Check all possible substrings for uniqueness.
    Most intuitive but least efficient approach.
    """
    def all_unique(substring: str) -> bool:
        return len(set(substring)) == len(substring)
    
    n = len(s)
    max_length = 0
    
    # Check all possible substrings
    for i in range(n):
        for j in range(i + 1, n + 1):
            if all_unique(s[i:j]):
                max_length = max(max_length, j - i)
    
    return max_length

# =============================================================================
# Approach 2: Sliding Window with Set
# Time: O(2n) = O(n), Space: O(min(m,n))
# =============================================================================

def lengthOfLongestSubstring_sliding_window(s: str) -> int:
    """
    Sliding Window: Use two pointers and a set to track unique characters.
    Expand right pointer, contract left when duplicates found.
    """
    char_set = set()
    left = 0
    max_length = 0
    
    for right in range(len(s)):
        # Shrink window from left until no duplicates
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        
        # Add current character and update max length
        char_set.add(s[right])
        max_length = max(max_length, right - left + 1)
    
    return max_length

# =============================================================================
# Approach 3: Optimized Sliding Window with Dictionary
# Time: O(n), Space: O(min(m,n))
# =============================================================================

def lengthOfLongestSubstring_optimized(s: str) -> int:
    """
    Optimized Sliding Window: Use dictionary to store character indices.
    Jump left pointer directly instead of incrementing one by one.
    """
    char_index = {}  # Character -> last seen index
    left = 0
    max_length = 0
    
    for right, char in enumerate(s):
        # If character seen and within current window, jump left pointer
        if char in char_index and char_index[char] >= left:
            left = char_index[char] + 1
        
        # Update character index and max length
        char_index[char] = right
        max_length = max(max_length, right - left + 1)
    
    return max_length

# =============================================================================
# Approach 4: Pythonic with dict.get()
# Time: O(n), Space: O(min(m,n))
# =============================================================================

def lengthOfLongestSubstring_pythonic(s: str) -> int:
    """
    Pythonic approach using dict.get() for cleaner code.
    """
    seen = {}
    left = max_length = 0
    
    for right, char in enumerate(s):
        # Use dict.get() with default -1 for unseen characters
        left = max(left, seen.get(char, -1) + 1)
        seen[char] = right
        max_length = max(max_length, right - left + 1)
    
    return max_length

# =============================================================================
# Approach 5: Using Deque (Educational)
# Time: O(n^2), Space: O(n)
# =============================================================================

def lengthOfLongestSubstring_deque(s: str) -> int:
    """
    Using deque to maintain current window of unique characters.
    Educational approach showing different data structure usage.
    """
    if not s:
        return 0
    
    max_length = 0
    window = deque()
    
    for char in s:
        # Remove characters from left until no duplicate
        while char in window:
            window.popleft()
        
        window.append(char)
        max_length = max(max_length, len(window))
    
    return max_length

# =============================================================================
# Approach 6: Functional Programming Style
# Time: O(n^2), Space: O(n)
# =============================================================================

def lengthOfLongestSubstring_functional(s: str) -> int:
    """
    Functional programming approach using generator expressions.
    Less efficient but demonstrates different programming paradigm.
    """
    if not s:
        return 0
    
    def expand_from_position(start: int) -> int:
        """Expand substring starting from given position."""
        seen = set()
        for i in range(start, len(s)):
            if s[i] in seen:
                break
            seen.add(s[i])
        return len(seen)
    
    return max(expand_from_position(i) for i in range(len(s)))

# =============================================================================
# Approach 7: Using List Comprehension (Advanced Python)
# Time: O(n^2), Space: O(n)
# =============================================================================

def lengthOfLongestSubstring_list_comprehension(s: str) -> int:
    """
    Advanced Python using list comprehension and itertools.
    Demonstrates Python's powerful built-in functions.
    """
    if not s:
        return 0
    
    # Generate all possible starting positions
    def unique_length_from(start: int) -> int:
        chars = []
        for char in s[start:]:
            if char in chars:
                break
            chars.append(char)
        return len(chars)
    
    return max([unique_length_from(i) for i in range(len(s))])

# =============================================================================
# Approach 8: Generator-based (Memory Efficient)
# Time: O(n^2), Space: O(k) where k is max unique chars
# =============================================================================

def lengthOfLongestSubstring_generator(s: str) -> int:
    """
    Generator-based approach for memory efficiency.
    Useful for very long strings where memory is a concern.
    """
    def generate_substrings():
        """Generator that yields lengths of unique substrings."""
        for i in range(len(s)):
            seen = set()
            length = 0
            for j in range(i, len(s)):
                if s[j] in seen:
                    break
                seen.add(s[j])
                length += 1
            yield length
    
    return max(generate_substrings()) if s else 0

# =============================================================================
# Approach 9: Recursive with Memoization
# Time: O(n^2), Space: O(n^2) for memoization
# =============================================================================

def lengthOfLongestSubstring_recursive_memo(s: str) -> int:
    """
    Recursive approach with memoization for educational purposes.
    Shows how to apply recursion to sliding window problems.
    """
    memo = {}
    
    def helper(start: int, seen: frozenset) -> int:
        if start >= len(s):
            return 0
        
        # Memoization key
        key = (start, seen)
        if key in memo:
            return memo[key]
        
        current_char = s[start]
        
        if current_char in seen:
            # Character already seen, can't extend current substring
            result = helper(start + 1, frozenset())
        else:
            # Two choices: include current char or start new substring
            include = 1 + helper(start + 1, seen | {current_char})
            exclude = helper(start + 1, frozenset())
            result = max(include, exclude)
        
        memo[key] = result
        return result
    
    return helper(0, frozenset()) if s else 0

# =============================================================================
# Helper Functions for Testing and Analysis
# =============================================================================

def find_longest_substring_details(s: str) -> Tuple[int, str, int, int]:
    """
    Return length, actual substring, start index, and end index.
    """
    if not s:
        return 0, "", 0, 0
    
    char_index = {}
    left = 0
    max_length = 0
    best_left = 0
    best_right = 0
    
    for right, char in enumerate(s):
        if char in char_index and char_index[char] >= left:
            left = char_index[char] + 1
        
        char_index[char] = right
        
        if right - left + 1 > max_length:
            max_length = right - left + 1
            best_left = left
            best_right = right
    
    substring = s[best_left:best_right + 1]
    return max_length, substring, best_left, best_right

def analyze_string_patterns(s: str) -> Dict[str, any]:
    """
    Analyze string to provide insights about character distribution.
    """
    if not s:
        return {"empty": True}
    
    char_count = defaultdict(int)
    for char in s:
        char_count[char] += 1
    
    return {
        "length": len(s),
        "unique_chars": len(char_count),
        "max_char_frequency": max(char_count.values()),
        "most_frequent_char": max(char_count.items(), key=lambda x: x[1]),
        "has_repeats": any(count > 1 for count in char_count.values()),
        "all_unique": len(char_count) == len(s)
    }

# =============================================================================
# Testing Framework
# =============================================================================

def test_all_approaches():
    """Comprehensive testing of all approaches."""
    print("=== Python Implementation: Longest Substring Without Repeating Characters ===\n")
    
    # Test cases from problem statement and edge cases
    test_cases = [
        "abcabcbb",    # Expected: 3 ("abc")
        "bbbbb",       # Expected: 1 ("b") 
        "pwwkew",      # Expected: 3 ("wke")
        "",            # Expected: 0
        "au",          # Expected: 2 ("au")
        "dvdf",        # Expected: 3 ("vdf")
        "abcdef",      # Expected: 6 ("abcdef")
        "aab",         # Expected: 2 ("ab")
        "tmmzuxt",     # Expected: 5 ("mzuxt")
        "abba",        # Expected: 2 ("ab" or "ba")
        "nfpdmpi",     # Expected: 5 ("nfpdm")
        "ggububgvfk",  # Expected: 6 ("gubvfk")
        "a",           # Single character
        "ab",          # Two different
        "aa",          # Two same
        " ",           # Space
        "!@#$%^&*()",  # Special characters
        "12321",       # Numbers
        "abcdefghijklmnopqrstuvwxyz"  # All alphabet
    ]
    
    # All approaches to test
    approaches = [
        ("Brute Force", lengthOfLongestSubstring_bruteforce),
        ("Sliding Window", lengthOfLongestSubstring_sliding_window),
        ("Optimized", lengthOfLongestSubstring_optimized),
        ("Pythonic", lengthOfLongestSubstring_pythonic),
        ("Deque", lengthOfLongestSubstring_deque),
        ("Functional", lengthOfLongestSubstring_functional),
        ("List Comprehension", lengthOfLongestSubstring_list_comprehension),
        ("Generator", lengthOfLongestSubstring_generator)
    ]
    
    print("Testing All Approaches:")
    print("======================\n")
    
    for i, test_string in enumerate(test_cases):
        print(f"Test {i + 1}: Input = \"{test_string}\"")
        
        # Get detailed analysis
        length, substring, start, end = find_longest_substring_details(test_string)
        analysis = analyze_string_patterns(test_string)
        
        print(f"  Expected result: Length = {length}, Substring = \"{substring}\" (indices {start}-{end})")
        
        # Test all approaches
        results = []
        for name, func in approaches:
            try:
                start_time = time.time()
                result = func(test_string)
                end_time = time.time()
                
                results.append(result)
                execution_time = (end_time - start_time) * 1000  # Convert to milliseconds
                print(f"  {name:18}: {result:2d} ({execution_time:.4f}ms)")
            except Exception as e:
                print(f"  {name:18}: ERROR - {e}")
                results.append(None)
        
        # Check consistency
        valid_results = [r for r in results if r is not None]
        if len(set(valid_results)) <= 1:
            print("  ✓ All approaches consistent")
        else:
            print("  ✗ Inconsistent results!")
        
        # String analysis
        if analysis.get("length", 0) > 0:
            print(f"  String analysis: {analysis['unique_chars']} unique chars, "
                  f"max frequency: {analysis['max_char_frequency']}")
        
        print()

def benchmark_performance():
    """Performance benchmarking with different string sizes."""
    print("Performance Benchmarking:")
    print("========================\n")
    
    # Create test strings of different sizes
    test_strings = {
        "Short (10 chars)": "abcdefghij",
        "Medium (100 chars)": "".join(chr(ord('a') + i % 26) for i in range(100)),
        "Long (1000 chars)": "".join(chr(ord('a') + i % 26) for i in range(1000)),
        "Worst case (1000 'a')": "a" * 1000
    }
    
    # Fast approaches for large strings
    fast_approaches = [
        ("Sliding Window", lengthOfLongestSubstring_sliding_window),
        ("Optimized", lengthOfLongestSubstring_optimized),
        ("Pythonic", lengthOfLongestSubstring_pythonic),
        ("Deque", lengthOfLongestSubstring_deque)
    ]
    
    for description, test_string in test_strings.items():
        print(f"{description}:")
        print("-" * len(description))
        
        for name, func in fast_approaches:
            start_time = time.time()
            
            # Run multiple iterations for better measurement
            iterations = 1000 if len(test_string) <= 100 else 100
            for _ in range(iterations):
                result = func(test_string)
            
            end_time = time.time()
            avg_time = (end_time - start_time) / iterations * 1000
            
            print(f"  {name:15}: Result={result:3d}, Avg Time={avg_time:.6f}ms")
        
        print()

def demonstrate_edge_cases():
    """Demonstrate handling of various edge cases."""
    print("Edge Case Analysis:")
    print("==================\n")
    
    edge_cases = [
        ("Empty string", ""),
        ("Single char", "a"),
        ("All same", "aaaaa"),
        ("All different", "abcde"),
        ("Palindrome", "racecar"),
        ("With spaces", "a b c d"),
        ("Unicode", "αβγδε"),
        ("Numbers", "123456789"),
        ("Special chars", "!@#$%^&*()"),
        ("Mixed", "Hello World! 123")
    ]
    
    for description, test_case in edge_cases:
        length, substring, _, _ = find_longest_substring_details(test_case)
        analysis = analyze_string_patterns(test_case)
        
        print(f"{description:15}: \"{test_case}\"")
        print(f"  Result: Length={length}, Substring=\"{substring}\"")
        
        if analysis.get("length", 0) > 0:
            print(f"  Analysis: {analysis['unique_chars']} unique chars, "
                  f"{'all unique' if analysis['all_unique'] else 'has repeats'}")
        
        print()

# =============================================================================
# Main Execution
# =============================================================================

if __name__ == "__main__":
    test_all_approaches()
    benchmark_performance()
    demonstrate_edge_cases()
    
    print("Algorithm Complexity Summary:")
    print("============================")
    print("1. Brute Force:         O(n³) time, O(min(m,n)) space")
    print("2. Sliding Window:      O(n) time, O(min(m,n)) space") 
    print("3. Optimized:           O(n) time, O(min(m,n)) space")
    print("4. Pythonic:            O(n) time, O(min(m,n)) space")
    print("5. Deque:               O(n²) time, O(n) space")
    print("6. Functional:          O(n²) time, O(n) space")
    print("7. List Comprehension:  O(n²) time, O(n) space")
    print("8. Generator:           O(n²) time, O(k) space")
    print("9. Recursive + Memo:    O(n²) time, O(n²) space")
    print("\nRecommended: Optimized or Pythonic for best performance")
    print("Educational: Other approaches show different programming paradigms")

File: string/longest_substring_without_repeating_characters/longest_substring_without_repeating_characters.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

/**
 * LeetCode #3: Longest Substring Without Repeating Characters
 * 
 * Problem: Given a string s, find the length of the longest substring 
 * without repeating characters.
 * 
 * Multiple implementations demonstrating different approaches and optimizations.
 */

// =============================================================================
// Approach 1: Brute Force - Check all substrings
// Time: O(n^3), Space: O(min(m,n)) where m is charset size
// =============================================================================

bool allUnique(char* s, int start, int end) {
    bool char_set[256] = {false};  // ASCII character set
    
    for (int i = start; i < end; i++) {
        unsigned char c = (unsigned char)s[i];
        if (char_set[c]) {
            return false;  // Duplicate found
        }
        char_set[c] = true;
    }
    return true;
}

int lengthOfLongestSubstring_BruteForce(char* s) {
    int n = strlen(s);
    if (n == 0) return 0;
    
    int max_len = 1;
    
    // Check all possible substrings
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j <= n; j++) {
            if (allUnique(s, i, j)) {
                int current_len = j - i;
                max_len = (current_len > max_len) ? current_len : max_len;
            }
        }
    }
    
    return max_len;
}

// =============================================================================
// Approach 2: Sliding Window with Character Set
// Time: O(2n) = O(n), Space: O(min(m,n))
// =============================================================================

int lengthOfLongestSubstring_SlidingWindow(char* s) {
    int n = strlen(s);
    if (n == 0) return 0;
    
    bool char_set[256] = {false};  // Track characters in current window
    int left = 0, right = 0;
    int max_len = 0;
    
    while (right < n) {
        unsigned char right_char = (unsigned char)s[right];
        
        // If character already in window, shrink from left
        while (char_set[right_char]) {
            char_set[(unsigned char)s[left]] = false;
            left++;
        }
        
        // Add current character to window
        char_set[right_char] = true;
        
        // Update maximum length
        int current_len = right - left + 1;
        max_len = (current_len > max_len) ? current_len : max_len;
        
        right++;
    }
    
    return max_len;
}

// =============================================================================
// Approach 3: Optimized Sliding Window with Hash Map
// Time: O(n), Space: O(min(m,n))
// 
// Key optimization: Instead of incrementing left one by one,
// jump directly to the position after the duplicate character.
// =============================================================================

int lengthOfLongestSubstring_Optimized(char* s) {
    int n = strlen(s);
    if (n == 0) return 0;
    
    int char_index[256];  // Store last seen index of each character
    for (int i = 0; i < 256; i++) {
        char_index[i] = -1;  // Initialize to -1 (not seen)
    }
    
    int left = 0;
    int max_len = 0;
    
    for (int right = 0; right < n; right++) {
        unsigned char current_char = (unsigned char)s[right];
        int last_seen = char_index[current_char];
        
        // If character was seen and is within current window
        if (last_seen >= left) {
            left = last_seen + 1;  // Jump left pointer
        }
        
        // Update last seen index
        char_index[current_char] = right;
        
        // Update maximum length
        int current_len = right - left + 1;
        max_len = (current_len > max_len) ? current_len : max_len;
    }
    
    return max_len;
}

// =============================================================================
// Approach 4: Space-Optimized for specific character sets
// For problems with known small character sets (e.g., lowercase letters only)
// Time: O(n), Space: O(1) for fixed charset
// =============================================================================

int lengthOfLongestSubstring_Lowercase(char* s) {
    int n = strlen(s);
    if (n == 0) return 0;
    
    int char_index[26];  // Only for lowercase letters a-z
    for (int i = 0; i < 26; i++) {
        char_index[i] = -1;
    }
    
    int left = 0, max_len = 0;
    
    for (int right = 0; right < n; right++) {
        char current_char = s[right];
        
        // Check if it's a lowercase letter
        if (current_char >= 'a' && current_char <= 'z') {
            int char_idx = current_char - 'a';
            
            if (char_index[char_idx] >= left) {
                left = char_index[char_idx] + 1;
            }
            
            char_index[char_idx] = right;
            
            int current_len = right - left + 1;
            max_len = (current_len > max_len) ? current_len : max_len;
        }
    }
    
    return max_len;
}

// =============================================================================
// Approach 5: Recursive Solution (Educational)
// Time: O(n^2), Space: O(n) for recursion stack
// =============================================================================

int lengthOfLongestSubstring_Recursive_Helper(char* s, int start, int end, 
                                            bool* seen, int max_len) {
    if (start > end) {
        return max_len;
    }
    
    // Try starting a new window from current position
    memset(seen, false, 256 * sizeof(bool));
    int local_len = 0;
    
    for (int i = start; i <= end; i++) {
        unsigned char c = (unsigned char)s[i];
        if (seen[c]) {
            break;
        }
        seen[c] = true;
        local_len++;
    }
    
    int new_max = (local_len > max_len) ? local_len : max_len;
    
    // Try starting from next position
    return lengthOfLongestSubstring_Recursive_Helper(s, start + 1, end, seen, new_max);
}

int lengthOfLongestSubstring_Recursive(char* s) {
    int n = strlen(s);
    if (n == 0) return 0;
    
    bool seen[256];
    return lengthOfLongestSubstring_Recursive_Helper(s, 0, n - 1, seen, 0);
}

// =============================================================================
// Helper Functions for Testing and Display
// =============================================================================

typedef struct {
    int length;
    int start_index;
    int end_index;
    char substring[256];
} SubstringResult;

SubstringResult findLongestSubstringWithDetails(char* s) {
    SubstringResult result = {0, 0, 0, ""};
    int n = strlen(s);
    
    if (n == 0) return result;
    
    int char_index[256];
    for (int i = 0; i < 256; i++) {
        char_index[i] = -1;
    }
    
    int left = 0;
    int max_len = 0;
    int best_left = 0, best_right = 0;
    
    for (int right = 0; right < n; right++) {
        unsigned char current_char = (unsigned char)s[right];
        
        if (char_index[current_char] >= left) {
            left = char_index[current_char] + 1;
        }
        
        char_index[current_char] = right;
        
        int current_len = right - left + 1;
        if (current_len > max_len) {
            max_len = current_len;
            best_left = left;
            best_right = right;
        }
    }
    
    result.length = max_len;
    result.start_index = best_left;
    result.end_index = best_right;
    
    // Copy substring
    int substring_len = best_right - best_left + 1;
    strncpy(result.substring, s + best_left, substring_len);
    result.substring[substring_len] = '\0';
    
    return result;
}

void printTestResult(char* input, int test_num) {
    printf("Test %d: Input = \"%s\"\n", test_num, input);
    
    // Get detailed result
    SubstringResult detail = findLongestSubstringWithDetails(input);
    
    // Test all approaches
    int bf_result = lengthOfLongestSubstring_BruteForce(input);
    int sw_result = lengthOfLongestSubstring_SlidingWindow(input);
    int opt_result = lengthOfLongestSubstring_Optimized(input);
    int rec_result = lengthOfLongestSubstring_Recursive(input);
    
    printf("  Brute Force:     %d\n", bf_result);
    printf("  Sliding Window:  %d\n", sw_result);
    printf("  Optimized:       %d\n", opt_result);
    printf("  Recursive:       %d\n", rec_result);
    printf("  Longest substring: \"%s\" (indices %d-%d)\n", 
           detail.substring, detail.start_index, detail.end_index);
    
    // Check consistency
    if (bf_result == sw_result && sw_result == opt_result && opt_result == rec_result) {
        printf("  ✓ All approaches consistent\n");
    } else {
        printf("  ✗ Inconsistent results!\n");
    }
    printf("\n");
}

// =============================================================================
// Performance Benchmarking
// =============================================================================

typedef struct {
    char* name;
    int (*func)(char*);
} Algorithm;

void benchmarkAlgorithms(char* test_string) {
    Algorithm algorithms[] = {
        {"Brute Force", lengthOfLongestSubstring_BruteForce},
        {"Sliding Window", lengthOfLongestSubstring_SlidingWindow},
        {"Optimized", lengthOfLongestSubstring_Optimized},
        {"Recursive", lengthOfLongestSubstring_Recursive}
    };
    
    int num_algorithms = sizeof(algorithms) / sizeof(algorithms[0]);
    int string_len = strlen(test_string);
    
    printf("Performance Benchmark (String length: %d)\n", string_len);
    printf("==========================================\n");
    
    for (int i = 0; i < num_algorithms; i++) {
        clock_t start = clock();
        
        // Run multiple times for better measurement
        int iterations = (string_len < 1000) ? 10000 : 100;
        int result = 0;
        
        for (int j = 0; j < iterations; j++) {
            result = algorithms[i].func(test_string);
        }
        
        clock_t end = clock();
        double time_taken = ((double)(end - start)) / CLOCKS_PER_SEC / iterations;
        
        printf("%-15s: Result=%d, Time=%.8f seconds\n", 
               algorithms[i].name, result, time_taken);
    }
    printf("\n");
}

// =============================================================================
// Main Testing Function
// =============================================================================

void runComprehensiveTests(void) {
    printf("=== LeetCode #3: Longest Substring Without Repeating Characters ===\n\n");
    
    // Basic test cases from problem statement
    char* test_cases[] = {
        "abcabcbb",    // Expected: 3 ("abc")
        "bbbbb",       // Expected: 1 ("b") 
        "pwwkew",      // Expected: 3 ("wke")
        "",            // Expected: 0
        "au",          // Expected: 2 ("au")
        "dvdf",        // Expected: 3 ("vdf")
        "abcdef",      // Expected: 6 ("abcdef")
        "aab",         // Expected: 2 ("ab")
        "tmmzuxt",     // Expected: 5 ("mzuxt")
        "abba",        // Expected: 2 ("ab" or "ba")
        "nfpdmpi",     // Expected: 5 ("nfpdm")
        "ggububgvfk"   // Expected: 6 ("gubvfk")
    };
    
    int num_tests = sizeof(test_cases) / sizeof(test_cases[0]);
    
    printf("Comprehensive Testing:\n");
    printf("=====================\n\n");
    
    for (int i = 0; i < num_tests; i++) {
        printTestResult(test_cases[i], i + 1);
    }
    
    // Edge case testing
    printf("Edge Case Testing:\n");
    printf("==================\n\n");
    
    char* edge_cases[] = {
        "a",           // Single character
        "ab",          // Two different characters
        "aa",          // Two same characters
        "abc",         // All different
        "aaa",         // All same
        " ",           // Space character
        "a b c",       // With spaces
        "!@#$%",       // Special characters
        "123321",      // Numbers with palindrome pattern
        "abcabcabcabc" // Long repeating pattern
    };
    
    int num_edge_cases = sizeof(edge_cases) / sizeof(edge_cases[0]);
    
    for (int i = 0; i < num_edge_cases; i++) {
        printTestResult(edge_cases[i], i + 1);
    }
    
    // Performance testing
    printf("Performance Analysis:\n");
    printf("====================\n\n");
    
    char short_string[] = "abcdef";
    char medium_string[] = "abcdefghijklmnopqrstuvwxyz";
    char long_string[1001];
    
    // Create a long string with pattern
    for (int i = 0; i < 1000; i++) {
        long_string[i] = 'a' + (i % 26);
    }
    long_string[1000] = '\0';
    
    benchmarkAlgorithms(short_string);
    benchmarkAlgorithms(medium_string);
    benchmarkAlgorithms(long_string);
    
    printf("Complexity Analysis:\n");
    printf("===================\n");
    printf("1. Brute Force:       O(n³) time, O(min(m,n)) space\n");
    printf("2. Sliding Window:    O(n) time, O(min(m,n)) space\n");
    printf("3. Optimized:         O(n) time, O(min(m,n)) space\n");
    printf("4. Recursive:         O(n²) time, O(n) space\n");
    printf("\nWhere n = string length, m = character set size\n");
    printf("For ASCII: m = 256, for lowercase letters: m = 26\n");
}

// =============================================================================
// Main Function
// =============================================================================

int main(void) {
    runComprehensiveTests();
    return 0;
}
